src/JsonSchema/SchemaStorage.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace JsonSchema;
+
+use JsonSchema\Entity\JsonPointer;
+use JsonSchema\Exception\UnresolvableJsonPointerException;
+use JsonSchema\Iterator\ObjectIterator;
+use JsonSchema\Uri\UriResolver;
+use JsonSchema\Uri\UriRetriever;
+
+class SchemaStorage
+{
+    protected $uriRetriever;
+    protected $uriResolver;
+    protected $schemas = array();
+
+    public function __construct(
+        UriRetrieverInterface $uriRetriever = null,
+        UriResolverInterface $uriResolver = null
+    )
+    {
+        $this->uriRetriever = $uriRetriever ?: new UriRetriever;
+        $this->uriResolver = $uriResolver ?: new UriResolver;
+    }
+
+    /**
+     * @return UriRetrieverInterface
+     */
+    public function getUriRetriever()
+    {
+        return $this->uriRetriever;
+    }
+
+    /**
+     * @return UriResolverInterface
+     */
+    public function getUriResolver()
+    {
+        return $this->uriResolver;
+    }
+
+    public function addSchema($id, $schema = null)
+    {
+        if (is_null($schema)) {
+            $schema = $this->uriRetriever->retrieve($id);
+        }
+        $objectIterator = new ObjectIterator($schema);
+        foreach ($objectIterator as $toResolveSchema) {
+            if (property_exists($toResolveSchema, '$ref') && is_string($toResolveSchema->{'$ref'})) {
```

Comment:
@FlorianSW 'id' and '$ref' are property names from jsonschema documentation, so I'm not sure we can rename them, not the '$ref' at least